File: branch_sums/branchSums.go
package main

import (
	"fmt"
	"io"
	"os"
)

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

type BST struct {
	Root *Node
}

func (node *Node) insert(value int) {
	if node == nil {
		return
	} else if value < node.Value {
		if node.Left == nil {
			node.Left = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Left.insert(value)
		}
	} else {
		if node.Right == nil {
			node.Right = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Right.insert(value)
		}
	}
}

func (tree *BST) insert(value int) *BST {
	if tree.Root == nil {
		tree.Root = &Node{Value: value, Left: nil, Right: nil}
	} else {
		tree.Root.insert(value)
	}
	return tree
}

func brunchSumsHelper(node *Node, sums *[]int, runSum int) {
	if node == nil {
		return
	}
	runSum += node.Value
	if (node.Left == nil) && (node.Right == nil) {
		fmt.Println(runSum)
		*sums = append(*sums, runSum)
		return
	}
	brunchSumsHelper(node.Left, sums, runSum)
	brunchSumsHelper(node.Right, sums, runSum)
}

func (tree *BST) brunchSums() []int {
	var sums []int
	brunchSumsHelper(tree.Root, &sums, 0)
	return sums
}

func print(w io.Writer, node *Node, ns int, ch rune) {
	if node == nil {
		return
	}
	for i := 0; i < ns; i++ {
		fmt.Fprint(w, " ")
	}
	fmt.Fprintf(w, "%c:%v\n", ch, node.Value)
	print(w, node.Left, ns+2, 'L')
	print(w, node.Right, ns+2, 'R')
}

func main() {
	tree := &BST{}
	tree.insert(5).
		insert(4).
		insert(3).
		insert(2).
		insert(1).
		insert(6).
		insert(7).
		insert(8).
		insert(9).
		insert(10)
	print(os.Stdout, tree.Root, 0, 'M')
	sums := []int{15, 45}
	fmt.Println(sums, "vs", tree.brunchSums())
}

File: branch_sums/branchSums.py
# Complexity O(n), Space O(n) n - number of nodes in tree
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def insert(self, values, i=0):
        if i >= len(values):
            return
        queue = [self]
        while len(queue) > 0:
            current = queue.pop(0)
            if current.left is None:
                current.left = BinaryTree(values[i])
                break
            queue.append(current.left)
            if current.right is None:
                current.right = BinaryTree(values[i])
                break
            queue.append(current.right)
        self.insert(values, i + 1)
        return self

def branchSumsHelper(node, sums, runSum):
    runSum += node.value
    if not node.left and not node.right:
        return sums.append(runSum)
    if node.left:
        branchSumsHelper(node.left, sums, runSum)
    if node.right:
        branchSumsHelper(node.right, sums, runSum)


def branchSums(root):
    sums = []
    branchSumsHelper(root, sums, 0)
    return sums


if __name__ == '__main__':
    tree = BinaryTree(1).insert([2, 3, 4, 5, 6, 7, 8, 9, 10])
    print("{} vs {}".format(branchSums(tree), [15, 16, 18, 10, 11]))
